examples/LM_Bundle_Adjust.ipynb

```diff
@@ -0,0 +1,725 @@
+      "\n",
+      "change in sigma0 of 2.3915314173450497e-12 converged!\n"
+     ]
+    }
+   ],
+   "source": [
+    "## sensors = generate_sensors(cubes) # generate sensors\n",
+    "cnet = io_controlnetwork.from_isis(network) # load in network\n",
+    "cnet = compute_apriori_ground_points(cnet, sensors) # calculate ground points\n",
+    "\n",
+    "### INPUTS ###\n",
+    "all_parameters = {sn: get_sensor_parameters(sensor) for sn, sensor in sensors.items()} #all parameters\n",
+    "parameters = {sn: parameter[6:12] for sn, parameter in all_parameters.items()} #just solving for camera angles and angle velocity\n",
+    "##############\n",
+    "\n",
+    "column_dict = compute_coefficient_columns(cnet, sensors, parameters)\n",
+    "num_parameters = max(col_range[1] for col_range in column_dict.values())\n",
+    "num_observations = 2 * len(cnet)\n",
+    "W_observations = np.eye(num_observations)\n",
+    "W_params = compute_parameter_weights(cnet, sensors, parameters, column_dict)\n",
+    "\n",
+    "iteration = 0\n",
+    "V = compute_residuals(cnet, sensors)\n",
+    "dX = np.zeros(W_params.shape[0]) #initialize for sigma calculatio\n",
+    "sigma0 = compute_sigma(V, dX, W_params, W_observations)\n",
+    "print(f'iteration {iteration}: sigma0 = {sigma0}\\n')\n",
+    "\n",
+    "max_iterations = 100\n",
+    "tol = 1e-10\n",
+    "total_correction = np.zeros(num_parameters)\n",
+    "damping = 0.001\n",
+    "for i in range(max_iterations):   \n",
+    "    iteration += 1\n",
+    "    old_sigma0 = sigma0\n",
+    "    \n",
+    "    J = compute_jacobian(cnet, sensors, parameters, column_dict)\n",
+    "    if i==0:\n",
+    "        damping *= np.max(J.T.dot(J))\n",
+    "    \n",
+    "    C = J.T.dot(W_observations).dot(V) - damping*np.identity(W_params.shape[0]).dot(total_correction) - W_params.dot(total_correction)\n",
+    "    N = J.T.dot(W_observations).dot(J) + damping*np.identity(W_params.shape[0]) + W_params\n",
+    "#    N = J.T.dot(W_observations).dot(J) + damping*W_params\n",
+    "#    C = J.T.dot(W_observations).dot(V) - damping*W_params.dot(total_correction) #different :D\n",
+    "    dX = np.linalg.inv(N).dot(C) # calculate change in camera parameters and ground points\n",
+    "    \n",
+    "\n",
+    "    total_correction += dX\n",
+    "    print(f'corrections: mean = {dX.mean()} min = {dX.min()} max = {dX.max()}')\n",
+    "    \n",
+    "    update_parameters(sensors, parameters, cnet, dX, column_dict)\n",
+    "    \n",
+    "    old_V = V\n",
+    "    V = compute_residuals(cnet, sensors)\n",
+    "    sigma0 = compute_sigma(V, dX, W_params, W_observations)\n",
+    "    if (sigma0 < old_sigma0):\n",
+    "        damping = max(1.0/3.0, (np.linalg.norm(old_V)**2 - np.linalg.norm(V)**2)/(dX.T.dot(damping*dX + C)))\n",
+    "        nu = 2\n",
+    "        print(\"DAMPING1:\", damping, \"NU1:\", nu)\n",
+    "    else:\n",
+    "        damping = damping * nu\n",
+    "        nu = 2 * nu\n",
```

Comment:
What happens if sigma increases and there is not a nu set? Should nu be set outside of the main loop?

Reply:
Probably.